docs/source/cpp/streaming_execution.rst

```diff
@@ -305,3 +305,601 @@ Datasets may be scanned multiple times; just make multiple scan
+        /*aggregates=*/{{"hash_count", &options}},
+        /*targets=*/{"a"},
+        /*names=*/{"count(a)"},
+        /*keys=*/{"b"}};
+
+    ARROW_ASSIGN_OR_RAISE(cp::ExecNode * aggregate,
+                              cp::MakeExecNode("aggregate", plan.get(), {source},
+                              aggregate_options));
+
+
+Scan-Node
+---------
+
+There is no class or struct defined as ScanNode in the source. 
+But :class:`arrow::compute::ScanNodeOptions` container includes the options
+passed to `MakeScanNode` internal function which creates an :class:`ExecNode`
+performing the defined task. This component includes a few options,
+defined in the :class:`arrow::compute::ScanNodeOptions` and this requires, 
+`std::shared_ptr<arrow::dataset::Dataset>`, 
+`std::shared_ptr<arrow::compute::ScanOptions>`, 
+`std::shared_ptr<arrow::util::AsyncToggle>`. 
+
+The :class:`arrow::compute::ScanOptions` includes the scaning options::
+
+    arrow::compute::Expression Materialize(std::vector<std::string> names,
+                              bool include_aug_fields = false) {
+        if (include_aug_fields) {
+            for (auto aug_name : {"__fragment_index",
+                "__batch_index", "__last_in_fragment"}) {
+            names.emplace_back(aug_name);
+            }
+        }
+
+        std::vector<arrow::compute::Expression> exprs;
+        for (const auto& name : names) {
+            exprs.push_back(arrow::compute::field_ref(name));
+        }
+
+        return arrow::compute::project(exprs, names);
+    }
+
+    auto options = std::make_shared<arrow::dataset::ScanOptions>();
+    // sync scanning is not supported by ScanNode
+    options->use_async = true;
+    options->projection = Materialize({});  // create empty projection
+
+    // construct the scan node
+    cp::ExecNode* scan;
+    auto scan_node_options = arrow::dataset::ScanNodeOptions{
+                              /*dataset*/dataset, 
+                              /*scan_options*/options};
+
+    ARROW_ASSIGN_OR_RAISE(scan,
+                            arrow::compute::MakeExecNode("scan", 
+                              plan.get(), 
+                              {}, 
+                              scan_node_options));
+
+``SinkNode``
+------------
+
+:class:`SinkNode` can be considered as the output or final node of an streaming 
+execution definition. :class:`arrow::compute::SinkNodeOptions` interface is used to pass 
+the required options. Requires 
+`std::function<arrow::Future<arrow::util::optional<arrow::compute::ExecBatch>>()>* generator`
+and `arrow::util::BackpressureOptions backpressure`. 
+
+Example::
+
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+    arrow::compute::ExecNode* sink;
+
+    ARROW_ASSIGN_OR_RAISE(sink, arrow::compute::MakeExecNode("sink", plan.get(), {source},
+                                                  arrow::compute::SinkNodeOptions{&sink_gen}));
+
+
+The output can be obtained as a table::
```

Comment:
Not for this PR but I wonder if we might want to someday just create a `to_table` sink node which hides the complexity of aysnc generators.  The user could just use a scan node as input and a ToTableNode as output and never be aware of any of that.

Reply:
I think it is a nice idea. It will be very useful for the high level C++ users.

Reply:
I would like to work on it, if we are moving for such a functionality.